### PR TITLE
fix:  class property and object key side effects detection

### PR DESCRIPTION
**Pull request by IWANABETHATGUY**

### CODE REVIEW

File: crates/rolldown/src/ast_scanner/side_effect_detector/mod.rs [hunk 1]
```diff
@@ -160,7 +181,13 @@ impl<'a> SideEffectDetector<'a> {
       }
       Expression::ClassExpression(cls) => self.detect_side_effect_of_class(cls),
       // Accessing global variables considered as side effect.
-      Expression::Identifier(ident) => self.is_unresolved_reference(ident),
+      Expression::Identifier(ident) => {
```
Comment (IWANABETHATGUY) [inline crates/rolldown/src/ast_scanner/side_effect_detector/mod.rs]: global variable `undefined` is considered as side effect free. https://esbuild.github.io/try/#YgAwLjIwLjIALS1idW5kbGUgLS1taW5pZnk9ZmFsc2UgLS1mb3JtYXQ9ZXNtAGUAZW50cnkuanMAY2xhc3MgVCB7CiAgW3VuZGVmaW5lZF0oKSB7fQp9AABmaWxlLmpzAHZhciB2ID0gdHlwZW9mIGxvY2FsVmFyCm1vZHVsZS5leHBvcnRzID0gdg

Comment (hyf0) [reply crates/rolldown/src/ast_scanner/side_effect_detector/mod.rs]: Should just use `detect_side_effect_of_indentifier` to unite the logic of detecting if a identifer have side effect. It's not just underfined but a list of identifiers.

---edited

It's better we just transform `member_expr.object` to expression and call `detect_side_effect_of_expression`.

File: crates/rolldown/src/ast_scanner/side_effect_detector/mod.rs [hunk 2]
```diff
@@ -23,6 +27,7 @@ static SIDE_EFFECT_FREE_MEMBER_EXPR_2: Lazy<FxHashSet<(&'static str, &'static st
       ("Object", "getOwnPropertyDescriptor"),
       ("Object", "getPrototypeOf"),
       ("Object", "getOwnPropertyNames"),
+      ("Symbol", "iterator"),
```
Comment (IWANABETHATGUY) [inline crates/rolldown/src/ast_scanner/side_effect_detector/mod.rs:30]: https://github.com/evanw/esbuild/blob/360d47230813e67d0312ad754cad2b6ee09b151b/internal/js_ast/js_ast_helpers.go#L2329-L2330

File: crates/rolldown/tests/esbuild/dce/tree_shaking_class_property/artifacts.snap [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+};
+let keep5 = class {
+	async [x]() {}
+};
+let keep6 = class {
+	[{toString() {}}] = 'x';
+};
+
+//#endregion
+```
```
Comment (IWANABETHATGUY) [inline crates/rolldown/tests/esbuild/dce/tree_shaking_class_property/artifacts.snap:33]:
```js
TestTreeShakingClassProperty
---------- /out.js ----------
let keep1 = class {
  [x] = "x";
};
let keep2 = class {
  [x]() {
  }
};
let keep3 = class {
  get [x]() {
  }
};
let keep4 = class {
  set [x](_) {
  }
};
let keep5 = class {
  async [x]() {
  }
};
let keep6 = class {
  [{ toString() {
  } }] = "x";
};
```

File: crates/rolldown/tests/esbuild/dce/tree_shaking_class_static_property/artifacts.snap [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+	static set [x](_) {}
+};
+let keep7 = class {
+	static async [x]() {}
+};
+let keep8 = class {
+	static [{toString() {}}] = 'x';
+};
+
+//#endregion
```
Comment (IWANABETHATGUY) [inline crates/rolldown/tests/esbuild/dce/tree_shaking_class_static_property/artifacts.snap:38]:
```js
TestTreeShakingClassStaticProperty
---------- /out.js ----------
let keep1 = class {
  static x = x;
};
let keep2 = class {
  static ["x"] = x;
};
let keep3 = class {
  static [x] = "x";
};
let keep4 = class {
  static [x]() {
  }
};
let keep5 = class {
  static get [x]() {
  }
};
let keep6 = class {
  static set [x](_) {
  }
};
let keep7 = class {
  static async [x]() {
  }
};
let keep8 = class {
  static [{ toString() {
  } }] = "x";
};
```

File: crates/rolldown/tests/esbuild/dce/dce_of_symbol_instances/artifacts.snap
Comment (IWANABETHATGUY) [inline crates/rolldown/tests/esbuild/dce/dce_of_symbol_instances/artifacts.snap:1]:
```js
TestDCEOfSymbolInstances
---------- /out/class.js ----------
// class.js
var Keep1 = class {
  *[Symbol.iterator]() {
  }
  [keep];
};
var Keep2 = class {
  [keep];
  *[Symbol.iterator]() {
  }
};
var Keep3 = class {
  *[Symbol.wtf]() {
  }
};

---------- /out/object.js ----------
// object.js
var keep1 = { *[Symbol.iterator]() {
}, [keep]: null };
var keep2 = { [keep]: null, *[Symbol.iterator]() {
} };
var keep3 = { *[Symbol.wtf]() {
} };
```

**Review by hyf0 — APPROVED** [review]

**Pull request merged**